### PR TITLE
Enforce that no cluster state can be published unless confirmed written to ZooKeeper

### PR DESCRIPTION
**Pull request by vekterli**
@geirst please review

This avoids some subtle edge cases in the first few seconds after leader
election, where we may inhibit storing a state to ZooKeeper but where it
may still be sent to the nodes once a grace period has expired.

To ensure liveness, add an extra condition where a cluster state version
may be published after the grace period has expired even if there are no
other triggers at this point.

### CODE REVIEW

**Comment by vekterli** [comment]
This might need a re-think. It looks like the code today could be susceptible to quorum disconnects in the ZK management thread. If so we'll have to handle this on a lower level.

**Comment by vekterli** [comment]
@hakonhall please also review. Have force-pushed a new version of the branch that
moves tracking of written versions all the way down to the `DatabaseHandler`.

This avoids a subtle edge case where the underlying ZK integration code
may fail silently a write, leaving the core controller logic to think that
it had actually durably persisted a particular state version.

In case of reelections racing with broadcasts, it would be possible for
leader-edge readbacks from ZK to retrieve a _lower_ version than one
that had already been published. This would cause the cluster controller
to get very confused about which cluster states nodes had already observed.
If a newly produced state version overlapped with a previously broadcast
state, the controller would not push the updated state to the nodes, as it
would (with good reason) assume the node had already observed it, seeing
that it had already ACKed the particular version number.

**Review by hakonhall — COMMENTED** [review]
Do we need to do anything special at the edge where a CC becomes master?  I'm concerned about a case where a version V is read from ZK when the CC becomes master, but if lastClusterStateVersionWrittenToZooKeeper is not set to V we may never broadcast V.

AFAIK the code ends up saving V back to ZK again, and therefore eventually trigger broadcast of V like we want.

**Comment by vekterli** [comment]
>Do we need to do anything special at the edge where a CC becomes master? I'm concerned about a case where a version V is read from ZK when the CC becomes master, but if lastClusterStateVersionWrittenToZooKeeper is not set to V we may never broadcast V.
>
>AFAIK the code ends up saving V back to ZK again, and therefore eventually trigger broadcast of V like we want.

Any post-election cluster state broadcast shall be preceded by a version number bump and subsequent write to ZooKeeper, so that should not be a concern (unless I've missed something).

File: clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/FleetController.java [hunk 1]
```diff
@@ -579,6 +583,9 @@ public void tick() throws Exception {
         synchronized (monitor) {
             boolean didWork;
             didWork = database.doNextZooKeeperTask(databaseContext);
+            // In case of delayed ZK writes caused by intermittent failures, make sure to tag any newly written versions.
+            systemStateBroadcaster.setLastClusterStateVersionWrittenToZooKeeper(
```
[the remark this answers is not in the document]
Comment (vekterli) [reply clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/FleetController.java:587]: Agreed, will change to explicitly pass the version instead.

**Comment by vekterli** [comment]
@hakonhall PTAL

**Review by hakonhall — APPROVED** [review]

**Review by geirst — COMMENTED** [review]
👍

**Pull request merged**